File: food-survivor/js/game.js
class Game {
    constructor() {
        this.canvas = document.getElementById('gameCanvas');
        this.ctx = this.canvas.getContext('2d');
        
        this.canvas.width = CONFIG.CANVAS_WIDTH;
        this.canvas.height = CONFIG.CANVAS_HEIGHT;
        
        // Create sprites first
        this.sprites = new Sprites();
        
        this.settings = new Settings();
        this.highScores = new HighScores();
        this.damageNumbers = [];
        
        // Initialize game state
        this.gameState = 'menu';
        this.isLooping = true;
        
        // Initialize game objects
        this.reset();
        
        // Setup logo
        const logoContainer = document.querySelector('.pixel-art-logo');
        if (logoContainer) {
            logoContainer.innerHTML = ''; // Clear any existing content
            logoContainer.appendChild(this.sprites.logo);
        }
        
        // Hide all overlays except main menu
        this.hideAllOverlays();
        document.getElementById('menu-screen').classList.remove('hidden');
        
        // Setup event listeners
        this.setupEventListeners();
        this.setupMenuListeners();
        
        // Start game loop
        requestAnimationFrame(() => this.gameLoop());
    }

    hideAllOverlays() {
        const overlays = [
            'menu-screen',
            'game-over-screen',
            'pause-menu',
            'settings-menu',
            'highscores-menu',
            'level-up-menu'
        ];
        
        overlays.forEach(id => {
            const element = document.getElementById(id);
            if (element) {
                element.classList.add('hidden');
            }
        });
    }

    showMenu() {
        this.gameState = 'menu';
        this.hideAllOverlays();
        document.getElementById('menu-screen').classList.remove('hidden');
    }

    startGame() {
        console.log('startGame called, current state:', this.gameState); // Debug log
        this.reset();
        this.gameState = 'playing';
        this.hideAllOverlays();
        
        // Initialize player and game objects
        this.player = new Player(CONFIG.CANVAS_WIDTH / 2, CONFIG.CANVAS_HEIGHT / 2);
        this.player.game = this;
        this.enemies = [];
        this.projectiles = [];
        this.particles = new ParticleSystem();
        
        console.log('Game started, state:', this.gameState); // Debug log
    }

    togglePause() {
        if (this.gameState === 'playing') {
            this.gameState = 'paused';
            document.getElementById('pause-menu').classList.remove('hidden');
        } else if (this.gameState === 'paused') {
            this.resumeGame();
        }
    }

    resumeGame() {
        this.gameState = 'playing';
        this.hideAllOverlays();
    }

    gameLoop() {
        if (this.gameState === 'playing') {
            this.update();
        }
        
        this.draw();
        requestAnimationFrame(() => this.gameLoop());
    }

    reset() {
        this.player = new Player(CONFIG.CANVAS_WIDTH / 2, CONFIG.CANVAS_HEIGHT / 2);
        this.player.game = this;
        this.enemies = [];
        this.projectiles = [];
        this.particles = new ParticleSystem();
        this.waveManager = new WaveManager(this);
        this.keys = {};
        this.mouseX = 0;
        this.mouseY = 0;
        this.lastShot = 0;
        this.lastAutoShot = 0;
        this.soundEnabled = false;
        this.score = 0;
        this.wave = 1;
        this.damageNumbers = [];
    }

    setupEventListeners() {
        window.addEventListener('keydown', (e) => {
            if (e.key === 'Escape') {
                if (this.gameState === 'playing') {
                    this.togglePause();
                } else if (this.gameState === 'paused') {
                    this.resumeGame();
                }
                return;
            }
            
            if (this.gameState === 'playing') {
                this.keys[e.key] = true;
                this.soundEnabled = true;
            }
        });

        window.addEventListener('keyup', (e) => {
            if (this.gameState === 'playing') {
                this.keys[e.key] = false;
            }
        });
        
        this.canvas.addEventListener('mousemove', (e) => {
            if (this.gameState === 'playing') {
                const rect = this.canvas.getBoundingClientRect();
                this.mouseX = e.clientX - rect.left;
                this.mouseY = e.clientY - rect.top;
            }
        });
    }

    setupMenuListeners() {
        const startButton = document.getElementById('start-button');
        if (startButton) {
            startButton.onclick = () => {
                console.log('Starting game...'); // Debug log
                this.startGame();
            };
        }

        const resumeButton = document.getElementById('resume-button');
        if (resumeButton) {
            resumeButton.onclick = () => this.resumeGame();
        }

        const pauseRestartButton = document.getElementById('pause-restart-button');
        if (pauseRestartButton) {
            pauseRestartButton.onclick = () => this.startGame();
        }

        const quitButton = document.getElementById('quit-to-menu');
        if (quitButton) {
            quitButton.onclick = () => {
                this.gameState = 'menu';
                this.hideAllOverlays();
                document.getElementById('menu-screen').classList.remove('hidden');
            };
        }
    }

    showSettings() {
        document.getElementById('menu-screen').classList.add('hidden');
        document.getElementById('settings-menu').classList.remove('hidden');
        
        // Apply current settings and setup listeners
        this.settings.applySettings();
        this.settings.setupListeners();
    }

    showHighScores() {
        document.getElementById('menu-screen').classList.add('hidden');
        document.getElementById('highscores-menu').classList.remove('hidden');
        this.highScores.updateDisplay();
    }

    update() {
        if (this.gameState !== 'playing') return;

        // Update game objects
        this.player.update(this.keys);
        this.enemies.forEach(enemy => enemy.update());
        this.projectiles = this.projectiles.filter(projectile => !projectile.update());
        this.particles.update();
        
        // Spawn enemies
        this.spawnEnemy();
        
        // Check collisions
        this.checkCollisions();
        
        // Auto-shoot
        this.autoShoot();
        
        // Update damage numbers
        this.damageNumbers = this.damageNumbers.filter(number => 
            number.update(1/60)
        );
    }

    draw() {
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        
        if (this.gameState === 'playing' || this.gameState === 'paused' || this.gameState === 'levelUp') {
            this.player.draw(this.ctx);
            this.enemies.forEach(enemy => enemy.draw(this.ctx));
            this.projectiles.forEach(projectile => projectile.draw(this.ctx));
            this.particles.draw(this.ctx);
            
            // Draw target indicator if enabled
            if (this.settings.settings.showTarget) {
                this.drawTargetIndicator();
            }
            
            // Draw damage numbers
            if (this.settings.settings.showDamage) {
                this.damageNumbers.forEach(number => number.draw(this.ctx));
            }
        }
    }

    drawTargetIndicator() {
        const target = this.findBestTarget();
        if (target) {
            this.ctx.save();
            this.ctx.strokeStyle = '#ffeb3b';
            this.ctx.lineWidth = 2;
            this.ctx.setLineDash([5, 5]);
            this.ctx.beginPath();
            this.ctx.arc(target.x, target.y, 20, 0, Math.PI * 2);
            this.ctx.stroke();
            this.ctx.restore();
        }
    }

    playSound(sound) {
        if (this.soundEnabled && ASSETS.SOUNDS[sound]) {
            try {
                ASSETS.SOUNDS[sound].volume = this.settings.settings.sfxVolume / 100;
                ASSETS.SOUNDS[sound].currentTime = 0;
                ASSETS.SOUNDS[sound].play().catch(e => {
                    if (e.name === 'NotAllowedError') {
                        this.soundEnabled = false;
                    }
                });
            } catch (e) {
                console.warn('Sound playback error:', e);
            }
        }
    }

    shoot() {
        const now = Date.now();
        if (now - this.lastShot >= this.player.weapons[0].fireRate) {
            this.projectiles.push(new Projectile(
                this.player.x + this.player.size / 2,
                this.player.y + this.player.size / 2,
                this.mouseX,
                this.mouseY,
                this.player.weapons[0]
            ));
            this.lastShot = now;
            this.playSound('shoot');
        }
    }

    spawnEnemy() {
        const now = Date.now();
        if (!this.lastEnemySpawn) {
            this.lastEnemySpawn = now;
        }
        
        if (now - this.lastEnemySpawn >= CONFIG.ENEMIES.SPAWN_RATE) {
            // Choose spawn position outside the screen
            let x, y;
            const side = Math.floor(Math.random() * 4); // 0: top, 1: right, 2: bottom, 3: left
            
            switch(side) {
                case 0: // top
                    x = Math.random() * CONFIG.CANVAS_WIDTH;
                    y = -32;
                    break;
                case 1: // right
                    x = CONFIG.CANVAS_WIDTH + 32;
                    y = Math.random() * CONFIG.CANVAS_HEIGHT;
                    break;
                case 2: // bottom
                    x = Math.random() * CONFIG.CANVAS_WIDTH;
                    y = CONFIG.CANVAS_HEIGHT + 32;
                    break;
                case 3: // left
                    x = -32;
                    y = Math.random() * CONFIG.CANVAS_HEIGHT;
                    break;
            }
            
            const randomType = CONFIG.ENEMIES.TYPES[
                Math.floor(Math.random() * CONFIG.ENEMIES.TYPES.length)
            ];
            
            const enemy = new Enemy(randomType, x, y, this.player);
            this.enemies.push(enemy);
            this.lastEnemySpawn = now;
        }
    }

    checkCollisions() {
        // Projectile-Enemy collisions
        this.projectiles = this.projectiles.filter(projectile => {
            let projectileHit = false;
            
            this.enemies = this.enemies.filter(enemy => {
                if (this.checkCollision(projectile, enemy)) {
                    const damage = projectile.weapon.damage;
                    const enemyDied = enemy.takeDamage(damage);
                    
                    // Add damage number
                    this.addDamageNumber(
                        enemy.x + enemy.size/2,
                        enemy.y,
                        damage
                    );
                    
                    this.particles.createHitEffect(
                        projectile.x,
                        projectile.y,
                        enemy.getEnemyColor()
                    );
                    projectileHit = true;
                    
                    if (enemyDied) {
                        this.player.addXp(enemy.xpValue);
                        this.addDamageNumber(
                            enemy.x + enemy.size/2,
                            enemy.y - 20,
                            enemy.xpValue,
                            'xp'
                        );
                        this.particles.createXPEffect(enemy.x, enemy.y);
                        return false;
                    }
                }
                return true;
            });
            
            return !projectileHit && !projectile.isExpired();
        });

        // Enemy-Player collisions
        this.enemies.forEach(enemy => {
            if (this.checkCollision(enemy, this.player)) {
                // Handle player damage with invincibility frames
                if (!this.player.isInvulnerable) {
                    this.player.takeDamage(enemy.damage);
                }
            }
        });
    }

    checkCollision(obj1, obj2) {
        return obj1.x < obj2.x + obj2.size &&
               obj1.x + obj1.size > obj2.x &&
               obj1.y < obj2.y + obj2.size &&
               obj1.y + obj1.size > obj2.y;
    }

    findBestTarget() {
        let bestTarget = null;
        let bestScore = -Infinity;
        
        for (const enemy of this.enemies) {
            const dx = enemy.x - this.player.x;
            const dy = enemy.y - this.player.y;
            const distance = Math.sqrt(dx * dx + dy * dy);
            
            // Calculate target score based on multiple factors
            const score = this.calculateTargetScore(enemy, distance);
            
            if (score > bestScore) {
                bestScore = score;
                bestTarget = enemy;
            }
        }
        
        return bestTarget;
    }

    calculateTargetScore(enemy, distance) {
        // Base score is inverse of distance (closer is better)
        let score = 1000 / (distance + 1);
        
        // Prioritize enemies with lower health
        score += (1 - enemy.health / enemy.maxHealth) * 500;
        
        // Prioritize enemies that are closer to the player
        if (distance < 200) {
            score += 300;
        }
        
        // Prioritize more dangerous enemies
        score += enemy.damage * 10;
        
        return score;
    }

    autoShoot() {
        const now = Date.now();
        if (!this.lastAutoShot) {
            this.lastAutoShot = now;
        }

        // Fire all weapons that are ready
        this.player.weapons.forEach(weapon => {
            if (now - weapon.lastFired >= weapon.fireRate) {
                // Skip targeting for orbital/shield weapons
                if (weapon.isOrbital || weapon.isShield) {
                    this.player.fireWeapon(weapon);
                    weapon.lastFired = now;
                    return;
                }

                const target = this.findBestTargetForWeapon(weapon);
                if (target) {
                    // Update player's target coordinates for weapon firing
                    this.player.targetX = target.x + target.size/2;
                    this.player.targetY = target.y + target.size/2;
                    
                    // Fire weapon
                    this.player.fireWeapon(weapon);
                    weapon.lastFired = now;
                }
            }
        });
    }

    findBestTargetForWeapon(weapon) {
        let bestTarget = null;
        let bestScore = -Infinity;
        
        for (const enemy of this.enemies) {
            const dx = enemy.x - this.player.x;
            const dy = enemy.y - this.player.y;
            const distance = Math.sqrt(dx * dx + dy * dy);
            
            // Skip if enemy is out of weapon range
            if (distance > weapon.range) continue;
            
            // Calculate weapon-specific target score
            const score = this.calculateWeaponTargetScore(weapon, enemy, distance);
            
            if (score > bestScore) {
                bestScore = score;
                bestTarget = enemy;
            }
        }
        
        return bestTarget;
    }

    calculateWeaponTargetScore(weapon, enemy, distance) {
        let score = 1000 / (distance + 1);
        
        // Spread weapons prefer grouped enemies
        if (weapon.spread) {
            const nearbyEnemies = this.enemies.filter(other => {
                const dx = other.x - enemy.x;
                const dy = other.y - enemy.y;
                return Math.sqrt(dx * dx + dy * dy) < 100;
            }).length;
            score += nearbyEnemies * 200;
        }
        
        // High damage weapons prefer high health enemies
        if (weapon.damage > 15) {
            score += enemy.health * 0.5;
        }
        
        // Fast weapons prefer low health enemies
        if (weapon.fireRate < 300) {
            score += (1 - enemy.health/enemy.maxHealth) * 300;
        }
        
        return score;
    }

    addDamageNumber(x, y, amount, type = 'damage') {
        if (this.settings.settings.showDamage) {
            this.damageNumbers.push(new DamageNumber(x, y, amount, type));
        }
    }

    showLevelUpMenu() {
        this.gameState = 'levelUp';
        document.getElementById('level-up-menu').classList.remove('hidden');
    }

    hideLevelUpMenu() {
        this.gameState = 'playing';
        document.getElementById('level-up-menu').classList.add('hidden');
    }
}

// Initialize game when window is fully loaded
window.addEventListener('load', () => {
    new Game();
}); 

File: food-survivor-cpp/src/Game.cpp
#include "Game.h"
#include <stdexcept>
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <random>

// Define PI if we can't use std::numbers
constexpr float PI = 3.14159265358979323846f;

Game::Game() : 
    window(sf::VideoMode(CONFIG::WINDOW_WIDTH, CONFIG::WINDOW_HEIGHT), "Food Survivor"),
    // Load font first
    font([]{
        sf::Font f;
        if (!f.loadFromFile("assets/fonts/pixel.ttf")) {
            throw std::runtime_error("Could not load font");
        }
        return f;
    }()),
    // Initialize UI with font
    ui(font),
    state(GameState::Menu),
    lastEnemySpawn(0.f),
    score(0),
    wave(1),
    difficulty(1.f),
    soundEnabled(true),
    soundVolume(70.f)
{
    // Initialize random seed
    std::srand(static_cast<unsigned int>(std::time(nullptr)));
    
    window.setFramerateLimit(60);
    
    // Initialize player
    player = std::make_shared<Player>(
        sf::Vector2f(CONFIG::WINDOW_WIDTH/2, CONFIG::WINDOW_HEIGHT/2)
    );
    
    // Setup UI elements
    setupUI();
    
    // Initialize other systems
    loadSounds();
}

void Game::setupUI() {
    // Setup score text
    scoreText.setFont(font);
    scoreText.setCharacterSize(24);
    scoreText.setFillColor(sf::Color::White);
    scoreText.setPosition(10, 10);
    
    // Setup level text
    levelText.setFont(font);
    levelText.setCharacterSize(24);
    levelText.setFillColor(sf::Color::White);
    levelText.setPosition(10, 40);
    
    // Setup wave text
    waveText.setFont(font);
    waveText.setCharacterSize(24);
    waveText.setFillColor(sf::Color::White);
    waveText.setPosition(10, 70);
    
    updateUI();
}

void Game::updateUI() {
    scoreText.setString("Score: " + std::to_string(score));
    levelText.setString("Level: " + std::to_string(player->getLevel()));
    waveText.setString("Wave: " + std::to_string(wave));
}

void Game::run() {
    sf::Clock clock;
    
    while (window.isOpen()) {
        float deltaTime = clock.restart().asSeconds();
        
        processEvents();
        update(deltaTime);
        render();
    }
}

void Game::processEvents() {
    sf::Event event;
    while (window.pollEvent(event)) {
        if (event.type == sf::Event::Closed) {
            window.close();
        }
        else if (event.type == sf::Event::KeyPressed) {
            switch (event.key.code) {
                case sf::Keyboard::Escape:
                    if (state == GameState::Playing) {
                        togglePause();
                    }
                    break;
                case sf::Keyboard::Space:
                    if (state == GameState::Menu || state == GameState::GameOver) {
                        startGame();
                    }
                    break;
            }
        }
        
        if (state == GameState::LevelUp) {
            sf::Vector2i mousePos = sf::Mouse::getPosition(window);
            sf::Vector2f worldPos = window.mapPixelToCoords(mousePos);
            
            if (event.type == sf::Event::MouseButtonPressed) {
                if (ui.handleClick(worldPos)) {
                    handleLevelUp(ui.getSelectedOption());
                }
            }
        }
    }

    if (state == GameState::Playing) {
        // Update input state
        keys[sf::Keyboard::W] = sf::Keyboard::isKeyPressed(sf::Keyboard::W);
        keys[sf::Keyboard::A] = sf::Keyboard::isKeyPressed(sf::Keyboard::A);
        keys[sf::Keyboard::S] = sf::Keyboard::isKeyPressed(sf::Keyboard::S);
        keys[sf::Keyboard::D] = sf::Keyboard::isKeyPressed(sf::Keyboard::D);
        
        // Update mouse position
        sf::Vector2i mousePixelPos = sf::Mouse::getPosition(window);
        mousePos = window.mapPixelToCoords(mousePixelPos);
    }
}

void Game::startGame() {
    state = GameState::Playing;
    score = 0;
    wave = 1;
    difficulty = 1.0f;
    
    // Reset player
    player = std::make_shared<Player>(
        sf::Vector2f(CONFIG::WINDOW_WIDTH/2, CONFIG::WINDOW_HEIGHT/2)
    );
    
    // Clear game objects
    enemies.clear();
    projectiles.clear();
    particles = ParticleSystem();
    
    // Reset timers
    lastEnemySpawn = gameClock.getElapsedTime().asSeconds();
    
    // Update UI
    updateUI();
}

void Game::update(float deltaTime) {
    if (state != GameState::Playing) {
        return;
    }
        
    // Update player
    player->update(deltaTime);
    
    // Handle auto-shooting
    autoShoot(deltaTime);
    
    // Update enemies
    for (auto& enemy : enemies) {
        enemy->update(deltaTime);
    }
    
    // Update projectiles
    auto projIt = projectiles.begin();
    while (projIt != projectiles.end()) {
        if ((*projIt)->update(deltaTime)) {
            projIt = projectiles.erase(projIt);
        } else {
            ++projIt;
        }
    }
    
    // Spawn enemies
    float currentTime = gameClock.getElapsedTime().asSeconds();
    if (currentTime - lastEnemySpawn >= CONFIG::ENEMIES::SPAWN_RATE / difficulty) {
        spawnEnemy();
        lastEnemySpawn = currentTime;
    }
    
    // Check collisions
    checkCollisions();
    
    // Update particles
    particles.update(deltaTime);
    
    // Update UI
    updateUI();
    
    // Update UI with current score
    ui.setScore(score);
}

void Game::autoShoot(float deltaTime) {
    // Get all player weapons
    const auto& weapons = player->getWeapons();
    
    for (const auto& weapon : weapons) {
        // Find best target for this weapon
        auto target = findBestTarget(*weapon);
        if (target && weapon->isReady()) {
            // Create projectile
            auto newProjectile = weapon->createProjectile(target->getPosition());
            if (newProjectile) {
                projectiles.push_back(newProjectile);
                // Play shoot sound
                playSound("shoot");
            }
        }
    }
}

std::shared_ptr<Enemy> Game::findBestTarget(const Weapon& weapon) {
    if (enemies.empty()) return nullptr;
    
    std::shared_ptr<Enemy> bestTarget = nullptr;
    float bestScore = -1.0f;
    
    for (const auto& enemy : enemies) {
        // Calculate distance to enemy
        sf::Vector2f diff = enemy->getPosition() - player->getPosition();
        float distance = std::hypot(diff.x, diff.y);
        
        // Skip if enemy is out of range
        if (distance > weapon.getRange()) continue;
        
        // Calculate target score
        float score = calculateTargetScore(*enemy, distance, weapon);
        
        // Update best target if this one is better
        if (score > bestScore) {
            bestScore = score;
            bestTarget = enemy;
        }
    }
    
    return bestTarget;
}

float Game::calculateTargetScore(const Enemy& enemy, float distance, const Weapon& weapon) {
    // Base score is inverse of distance (closer targets are better)
    float score = 1.0f / (distance + 1.0f);
    
    // Prioritize enemies with lower health
    score *= (1.0f + (1.0f - enemy.getHealth() / enemy.getMaxHealth()));
    
    // Prioritize enemies that are closer to the player
    if (distance < weapon.getRange() * 0.5f) {
        score *= 1.5f;
    }
    
    return score;
}

void Game::render() {
    window.clear(sf::Color(17, 17, 17));
    
    switch (state) {
        case GameState::Menu: {
            // Draw start screen
            sf::RectangleShape overlay(sf::Vector2f(CONFIG::WINDOW_WIDTH, CONFIG::WINDOW_HEIGHT));
            overlay.setFillColor(sf::Color(0, 0, 0, 200));
            window.draw(overlay);
            
            // Draw title
            sf::Text title("FOOD SURVIVOR", font);
            title.setCharacterSize(64);
            title.setFillColor(sf::Color::Yellow);
            title.setOutlineThickness(3.f);
            title.setOutlineColor(sf::Color(200, 150, 0));
            title.setOrigin(title.getLocalBounds().width/2.f, title.getLocalBounds().height/2.f);
            title.setPosition(CONFIG::WINDOW_WIDTH/2.f, CONFIG::WINDOW_HEIGHT/3.f);
            
            // Create start button
            sf::RectangleShape startButton(sf::Vector2f(200.f, 60.f));
            startButton.setOrigin(100.f, 30.f);
            startButton.setPosition(CONFIG::WINDOW_WIDTH/2.f, CONFIG::WINDOW_HEIGHT * 0.6f);
            startButton.setFillColor(sf::Color(50, 50, 50));
            startButton.setOutlineThickness(2.f);
            startButton.setOutlineColor(sf::Color::White);
            
            sf::Text startText("START GAME", font);
            startText.setCharacterSize(24);
            startText.setFillColor(sf::Color::White);
            startText.setOrigin(startText.getLocalBounds().width/2.f, startText.getLocalBounds().height/2.f);
            startText.setPosition(CONFIG::WINDOW_WIDTH/2.f, CONFIG::WINDOW_HEIGHT * 0.6f);
            
            // Check if mouse is hovering over start button
            sf::Vector2i mousePos = sf::Mouse::getPosition(window);
            if (startButton.getGlobalBounds().contains(mousePos.x, mousePos.y)) {
                startButton.setFillColor(sf::Color(80, 80, 80));
                startButton.setOutlineColor(sf::Color::Yellow);
                if (sf::Mouse::isButtonPressed(sf::Mouse::Left)) {
                    startGame();
                }
            }
            
            window.draw(title);
            window.draw(startButton);
            window.draw(startText);
            break;
        }
        
        case GameState::Playing: {
            // Draw game objects
            for (const auto& projectile : projectiles)
                projectile->draw(window);
                
            for (const auto& enemy : enemies)
                enemy->draw(window);
                
            player->draw(window);
            particles.draw(window);
            
            // Draw UI
            ui.draw(window, *player);
            break;
        }
        
        case GameState::Paused: {
            // Draw game state (frozen)
            for (const auto& projectile : projectiles)
                projectile->draw(window);
            for (const auto& enemy : enemies)
                enemy->draw(window);
            player->draw(window);
            particles.draw(window);
            
            // Draw UI even when paused
            ui.draw(window, *player);
            
            // Draw pause overlay
            sf::RectangleShape overlay(sf::Vector2f(CONFIG::WINDOW_WIDTH, CONFIG::WINDOW_HEIGHT));
            overlay.setFillColor(sf::Color(0, 0, 0, 128));
            window.draw(overlay);
            
            sf::Text pauseText("PAUSED", font);
            pauseText.setCharacterSize(48);
            pauseText.setFillColor(sf::Color::White);
            pauseText.setPosition(
                CONFIG::WINDOW_WIDTH/2.f - pauseText.getLocalBounds().width/2.f,
                CONFIG::WINDOW_HEIGHT/2.f - pauseText.getLocalBounds().height/2.f
            );
            window.draw(pauseText);
            break;
        }
        
        case GameState::GameOver: {
            sf::Text gameOverText("GAME OVER", font);
            gameOverText.setCharacterSize(48);
            gameOverText.setFillColor(sf::Color::Red);
            gameOverText.setPosition(
                CONFIG::WINDOW_WIDTH/2 - gameOverText.getLocalBounds().width/2,
                CONFIG::WINDOW_HEIGHT/3
            );
            
            sf::Text scoreText("Score: " + std::to_string(score), font);
            scoreText.setCharacterSize(24);
            scoreText.setFillColor(sf::Color::White);
            scoreText.setPosition(
                CONFIG::WINDOW_WIDTH/2 - scoreText.getLocalBounds().width/2,
                CONFIG::WINDOW_HEIGHT/2
            );
            
            sf::Text restartText("Press SPACE to restart", font);
            restartText.setCharacterSize(24);
            restartText.setFillColor(sf::Color::White);
            restartText.setPosition(
                CONFIG::WINDOW_WIDTH/2 - restartText.getLocalBounds().width/2,
                CONFIG::WINDOW_HEIGHT * 2/3
            );
            
            window.draw(gameOverText);
            window.draw(scoreText);
            window.draw(restartText);
            break;
        }
        
        case GameState::LevelUp: {
            // Draw game state behind level up menu
            for (const auto& projectile : projectiles)
                projectile->draw(window);
            for (const auto& enemy : enemies)
                enemy->draw(window);
            player->draw(window);
            particles.draw(window);
            
            // Draw UI including XP bar
            ui.draw(window, *player);
            break;
        }
    }
    
    window.display();
}

void Game::loadSounds() {
    // Load sound files
    std::vector<std::string> soundFiles = {
        "hit", "shoot", "levelup"
    };
    
    for (const auto& name : soundFiles) {
        Sound sound;
        std::string path = "assets/sounds/" + name + ".wav";
        if (sound.buffer.loadFromFile(path)) {
            sound.sound.setBuffer(sound.buffer);
            sound.sound.setVolume(soundVolume);
            sounds[name] = sound;
        } else {
            std::cerr << "Failed to load sound: " << path << std::endl;
        }
    }
}

void Game::spawnEnemy() {
    // Randomly select enemy type
    int typeIndex = rand() % CONFIG::ENEMIES::TYPES.size();
    const auto& enemyData = CONFIG::ENEMIES::TYPES[typeIndex];
    
    // Spawn position outside screen
    float angle = static_cast<float>(rand()) / RAND_MAX * 2 * PI;
    float distance = 100.f; // Spawn distance from screen edge
    
    float spawnX = CONFIG::WINDOW_WIDTH/2 + std::cos(angle) * (CONFIG::WINDOW_WIDTH/2 + distance);
    float spawnY = CONFIG::WINDOW_HEIGHT/2 + std::sin(angle) * (CONFIG::WINDOW_HEIGHT/2 + distance);
    
    // Create enemy
    auto enemy = std::make_shared<Enemy>(
        sf::Vector2f(spawnX, spawnY),
        enemyData,
        player.get()
    );
    
    enemies.push_back(enemy);
}

void Game::checkCollisions() {
    // Check projectile-enemy collisions
    for (auto projIt = projectiles.begin(); projIt != projectiles.end();) {
        bool projectileDestroyed = false;
        const auto& projectile = *projIt;
        
        for (auto enemyIt = enemies.begin(); enemyIt != enemies.end();) {
            const auto& enemy = *enemyIt;
            
            // Calculate collision
            sf::Vector2f diff = projectile->getPosition() - enemy->getPosition();
            float distance = std::hypot(diff.x, diff.y);
            float collisionRadius = (enemy->getSize() + projectile->getSize()) * 0.5f;
            
            if (distance < collisionRadius) {
                // Create hit particles at collision point
                particles.createHitEffect(
                    projectile->getPosition(),
                    projectile->getColor()
                );
                
                // Play hit sound
                playSound("hit");
                
                // Apply damage
                if (enemy->takeDamage(projectile->getDamage())) {
                    // Enemy died
                    int xpGained = enemy->getXPValue();
                    player->addXP(xpGained);
                    score += xpGained;
                    
                    // Create XP particle effect
                    particles.createXPEffect(
                        enemy->getPosition(),
                        xpGained
                    );
                    
                    // Create death explosion
                    particles.createExplosion(
                        enemy->getPosition(),
                        enemy->getColor(),
                        enemy->getSize() * 1.5f
                    );
                    
                    // Check if player leveled up
                    if (player->getXP() >= player->getNextLevelXP()) {
                        showLevelUpMenu();
                        playSound("levelup");
                        
                        // Create level up effect
                        particles.createLevelUpEffect(player->getPosition());
                    }
                    
                    enemyIt = enemies.erase(enemyIt);
                } else {
                    ++enemyIt;
                }
                
                // Destroy projectile unless it's piercing
                if (!projectile->isPiercing()) {
                    projectileDestroyed = true;
                    break;  // Exit enemy loop since projectile is destroyed
                }
            } else {
                ++enemyIt;
            }
        }
        
        // Remove destroyed projectile or move to next one
        if (projectileDestroyed) {
            projIt = projectiles.erase(projIt);
        } else {
            ++projIt;
        }
    }
    
    // Check enemy-player collisions
    for (const auto& enemy : enemies) {
        sf::Vector2f diff = player->getPosition() - enemy->getPosition();
        float distance = std::hypot(diff.x, diff.y);
        float collisionRadius = (player->getSize() + enemy->getSize()) * 0.5f;
        
        if (distance < collisionRadius) {
            player->takeDamage(enemy->getDamage());
            
            // Create hit effect
            particles.createHitEffect(
                player->getPosition() + diff * 0.5f,  // Create effect at collision point
                sf::Color::Red,
                20.f  // Larger effect for player hits
            );
            
            playSound("hit");
            
            if (player->getHealth() <= 0) {
                gameOver();
                break;
            }
        }
    }
}

void Game::gameOver() {
    state = GameState::GameOver;
    
    // Save high score if needed
    // TODO: Implement high score system
    
    // Clear game objects
    enemies.clear();
    projectiles.clear();
    particles = ParticleSystem();  // Reset particle system
    
    // Play game over sound
    playSound("gameover");
    
    // Update UI
    updateUI();
}

void Game::playSound(const std::string& name) {
    if (!soundEnabled) return;
    
    auto it = sounds.find(name);
    if (it != sounds.end()) {
        it->second.sound.play();
    }
}

void Game::togglePause() {
    if (state == GameState::Playing) {
        state = GameState::Paused;
        // Optional: Play pause sound
        // playSound("pause");
    } else if (state == GameState::Paused) {
        state = GameState::Playing;
        // Optional: Play unpause sound
        // playSound("unpause");
    }
}

void Game::showLevelUpMenu() {
    state = GameState::LevelUp;  // Set game state to level up
    
    // Pause the game while showing menu
    for (auto& enemy : enemies) {
        enemy->pause();
    }
    
    // Select 3 random powerups
    currentPowerupOptions.clear();
    std::vector<int> indices;
    for (size_t i = 0; i < CONFIG::POWERUPS::TYPES.size(); ++i) {
        indices.push_back(i);
    }
    
    // Shuffle indices
    std::random_device rd;
    std::mt19937 gen(rd());
    std::shuffle(indices.begin(), indices.end(), gen);
    
    // Take first 3 options
    for (int i = 0; i < std::min(3, (int)indices.size()); ++i) {
        currentPowerupOptions.push_back(CONFIG::POWERUPS::TYPES[indices[i]]);
    }
    
    // Show menu in UI
    ui.showLevelUpMenu(currentPowerupOptions);
}

void Game::handleLevelUp(int optionIndex) {
    const auto& powerup = currentPowerupOptions[optionIndex];
    
    // Apply powerup effect
    if (powerup.effect == "armor") {
        player->increaseMaxHealth(powerup.value);
    }
    else if (powerup.effect == "speed") {
        player->setSpeedMultiplier(powerup.value);
    }
    else if (powerup.effect == "damage") {
        player->setDamageMultiplier(powerup.value);
    }
    else if (powerup.effect == "bullet_speed") {
        player->setProjectileSpeedMultiplier(powerup.value);
    }
    else if (powerup.effect == "new_weapon") {
        // Find the weapon data that matches this powerup
        for (const auto& weaponData : CONFIG::WEAPONS::TYPES) {
            if (weaponData.name == powerup.name) {
                player->addWeapon(weaponData);
                break;
            }
        }
    }
    
    // Resume enemies
    for (auto& enemy : enemies) {
        enemy->resume();
    }
    
    ui.hideLevelUpMenu();
    state = GameState::Playing;
} 